File: atv1/Tabuleiro.java
import java.util.Arrays;

public class Tabuleiro{
    private int[][] matriz;
    private int tam;
    private int vazioLinha, vazioColuna;

    public Tabuleiro(int[] elementos){
        this.tam = (int) Math.sqrt(elementos.length);
        this.matriz = new int[tam][tam];
        int k = 0;
        //criação do tabuleiro
        for(int i = 0; i < tam; i++){
            for(int j = 0; j < tam; j++){
                matriz[i][j] = elementos[k++];
                if(matriz[i][j] == 0){
                    vazioLinha = i;
                    vazioColuna = j;
                }
            }
        }
    }

    public void mover(char movimento){
        int novaLinha = vazioLinha;
        int novaColuna = vazioColuna;
        // Escolha do movimento
        switch (movimento) {
            case 'u':
                novaLinha++;
                break;
            case 'd':
                novaLinha--;
                break;
            case 'l':
                novaColuna++;
                break;
            case 'r':
                novaColuna--;
                break;
            default:
                return;
        }
        //verificação se permanece dentro do tabuleiro
        if(novaLinha >= 0 && novaLinha < tam && novaColuna >=0 && novaColuna < tam){
            int temp = matriz[novaLinha][novaColuna];
            matriz[novaLinha][novaColuna] = 0;
            matriz[vazioLinha][vazioColuna] = temp;
            //atualização da posição do zero
            vazioLinha = novaLinha;
            vazioColuna = novaColuna;

        }
    }

    public boolean estaResolvido(){
        int correto = 0;
        for(int i = 0; i < tam; i++){
            for(int j = 0; j < tam; j++){
                if(matriz[i][j] != correto++) return false;
            }
        }
        return true;
    }

    public void imprimir(){
        String linhaDivisoria = "+";
        for(int i = 0; i < tam; i++) linhaDivisoria += "------+";

        System.out.println(linhaDivisoria);
        for(int i = 0; i < tam; i++){
            System.out.print("|");
            for(int j = 0; j < tam; j++){
                System.out.printf("%4d  |", matriz[i][j]);
            }
            System.out.println("\n" + linhaDivisoria);
        }
    }
}

File: atv2/tabuleiro.py
import math

class Tabuleiro:
    def __init__(self, elementos):
        self.tam = int(math.isqrt(len(elementos)))
        self.matriz = []
        self.vazio_linha = 0
        self.vazio_coluna = 0
        
        for i in range(self.tam):
            linha = []
            for j in range(self.tam):
                val = elementos[i * self.tam + j]
                linha.append(val)
                if val == 0:
                    self.vazio_coluna = j
                    self.vazio_linha = i
            self.matriz.append(linha)        
    
    def mover(self, movimento):
        nova_linha, nova_coluna = self.vazio_linha, self.vazio_coluna

        if movimento == 'u':
            nova_linha+=1
        elif movimento == 'd':
            nova_linha-=1
        elif movimento == 'l':
            nova_coluna+=1
        elif movimento == 'r':
            nova_coluna-=1
        else:
            return
        
        if 0 <= nova_linha < self.tam and 0 <= nova_coluna <self.tam:
            temp = self.matriz[nova_linha][nova_coluna]
            self.matriz[nova_linha][nova_coluna] = 0
            self.matriz[self.vazio_linha][self.vazio_coluna] = temp

            self.vazio_linha, self.vazio_coluna = nova_linha, nova_coluna

    def esta_resolvido(self):
        correto = 0
        for i in range(self.tam):
            for j in range(self.tam):
                if self.matriz[i][j] != correto:
                    return False
                correto += 1
        return True
    
    def imprimir(self):
        linha_divisoria = "+" + "------+" * self.tam
        print(linha_divisoria)

        for i in range(self.tam):
            print("|", end="")
            for j in range(self.tam):
                val = self.matriz[i][j]
                display = f"{val:4d}" if val != 0 else "    "
                print(f"{display}  |", end="")
            print("\n" + linha_divisoria)